### ISSUE TITLE
Allow Sequence and Index Files in Different Folders

### ISSUE THREAD

**Issue by DarioS** [PapenfussLab/gridss#195]
I typically have a folder named `sequences` with sub-folders such as hg38 for the FASTA files and another top-level folder named `indices` with sub-folders such as hg38 and bwa for the index of the reference genome. Would providing a user parameter such as `INDEX_FOLDER` allow such modularity to be maintained? Other software such as Bowtie and BWA allow these to be separated. For example,

> `bowtie` looks for the specified index first in the current directory, then in the `indexes` subdirectory under the directory where the `bowtie` executable is located, then looks in the directory specified in the `BOWTIE_INDEXES` environment variable.

In contrast, it seems that GRIDSS encourages aligner indices to clutter the reference sequence folder.

**Comment by d-cameron** [comment]
An additional `INDEX_FOLDER` adds additional complexity which is problematic for usability and aligner agnosticism. Your options are:

- Add a symlink to the reference itself into the index folder and use that. Somewhat annoying as you'd have to also create or symlink the .fai and .dict

- If you have a non-default index location then it is possible to specify it directly using the same mechanism that GRIDSS uses to support other aligners.
  - examples/gridss_separate.sh is an example script with the aligner command line is specified explicitly.

**Comment by d-cameron** [comment]
TLDR: the desired functionality can be achieved using the `gridss.SoftClipsToSplitRead` `ALIGNER_COMMAND_LINE` parameter.

**Comment by DarioS** [comment]
Thanks, but I can't run gridss at all. How should I change my command?

```
module load gridss/2.0.1
module load bwa/0.7.17
gridss.CallVariants INPUT=$INPUT OUTPUT=$SVpath REFERENCE_SEQUENCE=/project/StatBio/sequence/hg38/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna  BLACKLIST=/project/StatBio/problemRegions/hg38/ENCFF419RSJ.bed ALIGNER_COMMAND_LINE='bwa mem -t 2 /project/StatBio/indexes/bwa/hg38/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna %1$s' ASSEMBLY=$assemblyPath WORKING_DIR=/project/StatBio/structVars TMP_DIR=/scratch/StatBio WORKER_THREADS=2
-bash: gridss.CallVariants: command not found
```

I also tried `gridss` and `CallVariants` in place of `gridss.CallVariants`, but the command not found error still happens. I also wonder if `%1$s` should be in the command I use. I don't know what it does (`$s` might be some BASH variable but I don't see its definition in the script and `%1` might be a job number but it's just a guess) and I copied it from the example. GRIDSS is much harder for me to use than, for example, STAR or BWA, which don't require advanced shell scripting knowledge to do a biological analysis. For example, the Synopsis of bwa is `bwa mem ref.fa read1.fq read2.fq > aln-pe.sam ` which is easy to understand and immediately works for me.

**Comment by d-cameron** [comment]
> Thanks, but I can't run gridss at all. How should I change my command?

The easiest way is to look at the examples in https://github.com/PapenfussLab/gridss/tree/master/example and use the appropriate one.

By far the easiest solution is to create symlinks to the reference genome in the index directory and just use that:

```
ln -s /project/StatBio/sequence/hg38/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna /project/StatBio/indexes/bwa/hg38/GCA_000001405.15_GRCh38_no_alt_analysis_set.fa
samtools faidx /project/StatBio/indexes/bwa/hg38/GCA_000001405.15_GRCh38_no_alt_analysis_set.fa
```

Just take the `gridss.sh` example and adapt that to your usage.


If you really do need to use a different aligner, then the `gridss_separate.sh` example is the one you should adapt as it has an example of how to specify `ALIGNER_COMMAND_LINE`. The command line usage documentation on `ALIGNER_COMMAND_LINE` states the following:

```
Command line arguments to run external aligner. Aligner output should
be written to stdout and the records MUST match the input fastq order.
Java argument formatting is used with %1$s being the fastq file to align,
%2$s the reference genome, and %3$d the number of threads to use.
```

Your scenario where your index files are not in the same directory

>module load gridss/2.0.1

That module was created by someone local at your institution I'm not familiar what that module load actually does. GRIDSS is released as an all-in-one jar file so it uses `java` as the executable. Maybe it creates a GRIDSS_JAR environment variable?

>GRIDSS is much harder for me to use than, for example, STAR or BWA, which don't require advanced shell scripting knowledge to do a biological analysis.

This is true. There are actually 14 different step require to run GRIDSS. gridss.CallVariants wraps that all up and does that internally. The `gridss_separate.sh` example breaks those steps out so users can a) see what's going on under the hood (see the image on the github README), or make advanced changes such as changing the aligner used for realigning assembly contigs.

As for the syntax itself, I'm actually using the Picard tools command line processing tools so the command line user experience should be familiar to those that have used Picard tools. If you've never used picard tools then yes, the ARG=value syntax will be unfamiliar.

If you can use docker in your institution, there is also a GRIDSS docker image that is compatible with singularity.

Assuming you symlinked the reference genome, your command line was almost correct: it should look like the following:
```
java -ea -Xmx16g \
	-Dsamjdk.create_index=true \
	-Dsamjdk.use_async_io_read_samtools=true \
	-Dsamjdk.use_async_io_write_samtools=true \
	-Dsamjdk.use_async_io_write_tribble=true \
	-Dgridss.gridss.output_to_temp_file=true \
	-cp gridss-2.1.1-gridss-jar-with-dependencies.jar \
	gridss.CallVariants \
	INPUT=$INPUT \
	OUTPUT=$SVpath \
	REFERENCE_SEQUENCE=/project/StatBio/indexes/bwa/hg38/GCA_000001405.15_GRCh38_no_alt_analysis_set.fa \
	BLACKLIST=/project/StatBio/problemRegions/hg38/ENCFF419RSJ.bed \
	ASSEMBLY=$assemblyPath \
	WORKING_DIR=/project/StatBio/structVars \
	TMP_DIR=/scratch/StatBio \
	WORKER_THREADS=2
```

**Comment by DarioS** [comment]
This approach doesn't work with the `gridss.sh` wrapper in version 2.4.0 which uses `reference=$(readlink -f $reference || echo -n)` By following the link, it goes to the folder which lacks the bwa indexes and spends time indexing the FASTA file because of `` echo "Please create using `bwa index $reference`" 1>&2 ``

Note that the backticks cause the indexing code to actually be executed rather than printed to the log file and the progress message in the log appears truncated as 

> Please create using
